### PR TITLE
[Infra-cookbook] Fixed content tab data not loading on cookbook details page

### PR DESCRIPTION
**Pull request by vinay033**
Signed-off-by: Vinay Sharma <vsharma@chef.io>

### :nut_and_bolt: Description: What code changed, and why?

On the cookbook details page sometimes the content tab data is not loading, the cookbook’s content tab just shows an endless spinner.
### :chains: Related Resources
https://github.com/chef/automate/issues/4899
### :+1: Definition of Done
I have added changes for this now it will always work if sub-menus are not present.
### :athletic_shoe: How to Build and Test the Change
```
STEP 1
inside the hab studio

[default:/src:0]# build components/automate-ui-devproxy/
[default:/src:0]# start_automate_ui_background
[default:/src:0]# start_all_services

STEP 2
open new window
go to automate UI path

$ cd components/automate-ui
and run the command 

npm run serve:hab

navigate to `https://a2-dev.test/infrastructure/chef-servers`
add some sample data and then test the changes using tab change
```
To add data https://github.com/chef/automate/blob/master/dev-docs/adding-data/adding_test_data.md#adding-data-to-infra-views
```
Go to the `infrastructure tab` from the top navigation bar then you can see the `Chef Server`
If you have data then you can see the list of servers --> then click to any of the server --> see the list of orgs 
--> click to any of org ---> you can see the multiple tabs --> default tab is cookbooks tab --> 
you can see the list of cookbooks ---> click to any of specific cookbook (if attributes sub-menus are not present) then you can reproduce this issue.
```
### :white_check_mark: Checklist

**All PRs** must tick these:

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [x] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [x] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [x] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [x] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [x] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [x] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [ ] Tests added/updated? (all new code needs new tests)
- [ ] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### :camera: Screenshots, if applicable

https://user-images.githubusercontent.com/12297653/113917790-454c1200-97ff-11eb-8e3d-8b8436883e5a.mp4

### CODE REVIEW

File: components/automate-ui/src/app/modules/infra-proxy/cookbook-details/cookbook-details.component.ts [hunk 1]
```diff
@@ -138,6 +139,15 @@ export class CookbookDetailsComponent implements OnInit, OnDestroy {
...
           // check submenu exists and show default item details
+          for (const menu of this.menuList) {
+            if ( menu.subMenu.length ) {
+              this.defaultContent = menu.subMenu[0];
+              this.getContent(this.defaultContent);
+              this.menuIndex = this.menuList.indexOf(menu);
+              break;
+            }
+          }
+
```
Comment (SEAjamieD) [inline components/automate-ui/src/app/modules/infra-proxy/cookbook-details/cookbook-details.component.ts:150]: Suggest chaining and arrow functions when possible in keeping with current Chef convention
```
const submenus = this.menuList.filter(menu => menu.subMenu.length);

submenus.forEach(menu => {
    this.defaultContent = menu.subMenu[0];
    this.getContent(this.defaultContent);
    this.menuIndex = this.menuList.indexOf(menu);
});
```

Comment (vinay033) [reply components/automate-ui/src/app/modules/infra-proxy/cookbook-details/cookbook-details.component.ts:150]: Thank you for review @SEAjamieD 
Actually, your suggested code of block targeting the last menu, and our requirement is to target the first menu which has contained some submenu.
after we got the first menu (which has some submenu) I am using `break` to stop executing the remaining menus.

or we can use
```
const menus = this.menuList.filter(menu => menu.subMenu.length);
if ( menus.length ) {
   this.defaultContent = menus[0].subMenu[0];
   this.getContent(this.defaultContent);
   this.menuIndex = this.menuList.indexOf(menus[0]);
}
```

Chef Automate (13)

Comment (SEAjamieD) [reply components/automate-ui/src/app/modules/infra-proxy/cookbook-details/cookbook-details.component.ts:150]: Good follow through - thanks @vinay033.  My suggestion would be to adjust in the most appropriate way that allows us to still follow prior conventions, so if you can do this with arrow functions and chains, that would be ideal.

Comment (vinay033) [reply components/automate-ui/src/app/modules/infra-proxy/cookbook-details/cookbook-details.component.ts]: I have updated the PR, please check it once.

**Comment by SEAjamieD** [comment]
@vinay033 

Looking through the failing test led me to a solution that looks like this:
```
  it('Remote HTTP call when README file URL exists', () => {
    store.dispatch(new GetCookbookDetailsSuccess(cookbookDetails));
    fixture.detectChanges();

    const fileUrl = encodeURIComponent(cookbookDetails.root_files[0].url);
    const expectedUrl = `${env.infra_proxy_url}/servers/${server_id}/orgs/${org_id}/cookbooks/${cookbook_name}/${current_version}/file-content?url=${fileUrl}`;

    const req = httpMock.match(expectedUrl)[1];

    expect(req.request.method).toBe('GET');
    req.flush(readme_content);
    httpMock.verify();

    expect(component.readFileContent).toEqual(readme_content);
  });
```

The test above passes, but it brings up some questions around is this an issue in the testing setup or the components/effects/etc.  Looking through the component files, I don't immediately see any form of an error, so this just may be how the testing has to be.

It looks like by design, we need to use `httpMock.match()` because `httpMock.expectOne()` will fail when using parameters.  The large difference between `.match()` and `.expectOne()` is that match returns a TestRequest _Array_ instead of just a TestRequest item because it doesn't care about how many calls are made. https://angular.io/api/common/http/testing/HttpTestingController#match
Good info here too: https://stackoverflow.com/questions/49430213/error-expected-no-open-requests-found-1

if I place an `httpMock.verify()` immediately after `fixture.detectChanges()`, we get an error letting us know that there are actually _two_ calls being made.  This is where we'll need to dig in further to figure out if this is a test issue or a component issue.

**Comment by vinay033** [comment]
> it('Remote HTTP call when README file URL exists', () => {
>     store.dispatch(new GetCookbookDetailsSuccess(cookbookDetails));
>     fixture.detectChanges();
> 
>     const fileUrl = encodeURIComponent(cookbookDetails.root_files[0].url);
>     const expectedUrl = `${env.infra_proxy_url}/servers/${server_id}/orgs/${org_id}/cookbooks/${cookbook_name}/${current_version}/file-content?url=${fileUrl}`;
> 
>     const req = httpMock.match(expectedUrl)[1];
> 
>     expect(req.request.method).toBe('GET');
>     req.flush(readme_content);
>     httpMock.verify();
> 
>     expect(component.readFileContent).toEqual(readme_content);
>   });

Thank you so much @SEAjamieD 
for the information, right now I have replaced the test with this.

**Review by SEAjamieD — APPROVED** [review]
🎉

**Comment by SEAjamieD** [comment]
Out of scope of this ticket, so I've approved but let's make sure we get another ticket open to make sure that all of the selectable content inside the cookbooks tab is keyboard accessible.  Its super important we're doing our best to make sure everyone who may be using our products is able to use in the way that suits them best.

**Review by himanshi-chhabra — APPROVED** [review]

**Pull request merged**